nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/OAuth2TokenProviderImpl.java

```diff
@@ -0,0 +1,138 @@
+package org.apache.nifi.oauth2;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import okhttp3.FormBody;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.apache.nifi.annotation.lifecycle.OnEnabled;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.controller.AbstractControllerService;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.ssl.SSLContextService;
+import org.apache.nifi.util.StringUtils;
+
+import javax.net.ssl.SSLContext;
+import java.io.IOException;
+import java.util.List;
+
+import static org.apache.nifi.oauth2.Util.parseTokenResponse;
+
+public class OAuth2TokenProviderImpl extends AbstractControllerService implements OAuth2TokenProvider {
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    private String resourceServerUrl;
+    private ObjectMapper mapper = new ObjectMapper();
+    private SSLContext sslContext;
+    private SSLContextService sslContextService;
+
+    @OnEnabled
+    public void onEnabled(ConfigurationContext context) {
+        resourceServerUrl = context.getProperty(ACCESS_TOKEN_URL).evaluateAttributeExpressions().getValue();
+
+        sslContextService = context.getProperty(SSL_CONTEXT).asControllerService(SSLContextService.class);
+
+        sslContext = sslContextService == null ? null : sslContextService.createSSLContext(SSLContextService.ClientAuth.NONE);
+    }
+
+
+    @Override
+    public AccessToken getAccessTokenByPassword(String clientId, String clientSecret,
+                                                String username, String password) {
+        OkHttpClient.Builder builder = getClientBuilder();
+        OkHttpClient client = builder.build();
+
+        RequestBody body = new FormBody.Builder()
+                .add("username", username)
+                .add("password", password)
+                .add("client_id", clientId)
+                .add("client_secret", clientSecret)
+                .add("grant_type", "password")
+                .build();
+
+        Request newRequest = new Request.Builder()
+                .url(resourceServerUrl)
+                .post(body)
+                .build();
+
+        return executePost(client, newRequest);
+    }
+
+    private OkHttpClient.Builder getClientBuilder() {
```

Comment:
Could the client builder be created a single time when `onEnabled` is invoked? Otherwise it seems like this will be called several times and several OkHttpClient.Builder objects created/recreated when really only a single one is needed.

Maybe they go stale? I don't know just want to throw it out there.